File: email-processor/process_emails.py
#!/usr/bin/env python3

import os
import glob
from pathlib import Path
import extract_msg
from datetime import datetime


def get_msg_files(inbox_path):
    """Get all .msg files from the specified directory"""
    if not os.path.exists(inbox_path):
        print(f"Error: Directory {inbox_path} does not exist")
        return None
    
    pattern = os.path.join(inbox_path, "*.msg")
    msg_files = glob.glob(pattern)
    return msg_files


def parse_message_content(body):
    """Parse message content and return a map of question responses"""
    questionnaire = None
    homework = None
    coding_analysis = None
    
    if not body:
        return {
            "questionnaire": questionnaire,
            "homework": homework, 
            "coding_analysis": coding_analysis
        }
    
    lines = body.split('\n')
    for line in lines:
        line = line.strip().lower()
        if 'questionnaire:' in line:
            questionnaire = 'yes' in line
        if 'asked about homework:' in line:
            homework = 'yes' in line
        if 'did coding analysis:' in line:
            coding_analysis = 'yes' in line
    
    return {
        "questionnaire": questionnaire,
        "homework": homework,
        "coding_analysis": coding_analysis
    }


def process_single_email(file_path):
    """Process a single email message and extract basic info"""
    try:
        msg = extract_msg.Message(file_path)
        
        sender = msg.sender or "unknown"
        received_time = msg.date or datetime.fromtimestamp(os.path.getmtime(file_path))
        body = msg.body or ""
        
        parsed_data = parse_message_content(body)
        
        print(f"File: {os.path.basename(file_path)}")
        print(f"From: {sender}")
        print(f"Date: {received_time.strftime('%Y-%m-%d %H:%M:%S')}")
        print(f"Questionnaire: {parsed_data['questionnaire']}")
        print(f"Asked about homework: {parsed_data['homework']}")
        print(f"Did coding analysis: {parsed_data['coding_analysis']}")
        print("---")
        
        msg.close()
        return True
        
    except Exception as e:
        print(f"ERROR processing {os.path.basename(file_path)}: {e}")
        return False


def main():
    inbox_path = "../sample_emails"
    
    msg_files = get_msg_files(inbox_path)
    if msg_files is None:
        return 1
    
    msg_count = len(msg_files)
    print(f"Found {msg_count} .msg files in {inbox_path}\n")
    
    for file_path in msg_files:
        process_single_email(file_path)
    
    return 0


if __name__ == "__main__":
    exit(main())

File: process_emails.R
library(RDCOMClient)
library(DBI)
library(RSQLite)

# Configuration
INBOX_PATH <- "C:/EmailProcessing/inbox"
PROCESSED_PATH <- "C:/EmailProcessing/processed"
DB_PATH <- "C:/EmailProcessing/pcit_data.sqlite"

# Create directories if they don't exist
if (!dir.exists(PROCESSED_PATH)) {
  dir.create(PROCESSED_PATH, recursive = TRUE)
}

# Initialize SQLite database
init_database <- function() {
  con <- dbConnect(RSQLite::SQLite(), DB_PATH)
  
  # Create tables if they don't exist
  dbExecute(con, "
    CREATE TABLE IF NOT EXISTS email_data (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      clinic_email TEXT,
      processed_date TEXT,
      questionnaire TEXT,
      asked_homework TEXT,
      coding_analysis TEXT,
      created_at DATETIME DEFAULT CURRENT_TIMESTAMP
    )
  ")
  
  dbExecute(con, "
    CREATE TABLE IF NOT EXISTS clinic_lookup (
      email TEXT PRIMARY KEY,
      clinic_name TEXT
    )
  ")
  
  dbDisconnect(con)
}

# Parse email body for key-value pairs
parse_email_body <- function(body_text) {
  # Extract key-value pairs from body
  questionnaire <- if(grepl("Questionnaire:\\s*yes", body_text, ignore.case = TRUE)) "yes" else "no"
  homework <- if(grepl("Asked about homework:\\s*yes", body_text, ignore.case = TRUE)) "yes" else "no"
  coding <- if(grepl("Did coding analysis:\\s*yes", body_text, ignore.case = TRUE)) "yes" else "no"
  
  return(list(
    questionnaire = questionnaire,
    asked_homework = homework,
    coding_analysis = coding
  ))
}

# Process a single .msg file
process_msg_file <- function(file_path) {
  tryCatch({
    # Create Outlook application object
    outlook <- COMCreate("Outlook.Application")
    
    # Open the .msg file
    msg <- outlook$CreateItemFromTemplate(file_path)
    
    # Extract email data
    sender <- msg[["SenderEmailAddress"]]
    subject <- msg[["Subject"]]
    body <- msg[["Body"]]
    received_time <- msg[["ReceivedTime"]]
    
    cat("Processing email from:", sender, "\n")
    
    # Parse body content
    parsed_data <- parse_email_body(body)
    
    # Connect to database
    con <- dbConnect(RSQLite::SQLite(), DB_PATH)
    
    # Check if sender is in clinic lookup
    clinic_exists <- dbGetQuery(con, "SELECT COUNT(*) as count FROM clinic_lookup WHERE email = ?", 
                               params = list(sender))$count > 0
    
    if (clinic_exists) {
      # Insert data
      dbExecute(con, "
        INSERT INTO email_data (clinic_email, processed_date, questionnaire, asked_homework, coding_analysis)
        VALUES (?, ?, ?, ?, ?)
      ", params = list(
        sender,
        as.character(Sys.Date()),
        parsed_data$questionnaire,
        parsed_data$asked_homework,
        parsed_data$coding_analysis
      ))
      
      cat("✓ Data inserted successfully\n")
    } else {
      cat("⚠ Warning: Sender", sender, "not found in clinic lookup table\n")
    }
    
    dbDisconnect(con)
    
    # Move processed file
    file_name <- basename(file_path)
    processed_file <- file.path(PROCESSED_PATH, file_name)
    file.rename(file_path, processed_file)
    
    cat("✓ File moved to processed folder\n")
    
  }, error = function(e) {
    cat("✗ Error processing", file_path, ":", e$message, "\n")
  })
}

# Main processing function
process_emails <- function() {
  cat("Starting email processing...\n")
  
  # Get all .msg files in inbox
  msg_files <- list.files(INBOX_PATH, pattern = "\\.msg$", full.names = TRUE)
  
  if (length(msg_files) == 0) {
    cat("No .msg files found in inbox\n")
    return()
  }
  
  cat("Found", length(msg_files), ".msg files to process\n")
  
  # Process each file
  for (file_path in msg_files) {
    cat("\n--- Processing:", basename(file_path), "---\n")
    process_msg_file(file_path)
  }
  
  cat("\nEmail processing completed!\n")
}

# Initialize database on first run
init_database()

# Run the email processing
process_emails()